File: backend/src/main/java/com/blockchain/EHR/controller/FabricController.java
package com.blockchain.EHR.controller;

import com.blockchain.EHR.jwt.CustomUsernamePasswordAuthenticationToken;
import com.blockchain.EHR.jwt.JwtUtils;
import com.blockchain.EHR.model.EhrDocument;
import com.blockchain.EHR.model.LoginRequest;
import com.blockchain.EHR.model.UserEntity;
import com.blockchain.EHR.services.EhrService;
import com.blockchain.EHR.services.FabricService;
import com.blockchain.EHR.services.FabricUserRegistration;
import com.blockchain.EHR.services.UserInfoService;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

@Slf4j
@RestController
@RequestMapping("/fabric")
@AllArgsConstructor
public class FabricController {
    private final EhrService ehrService;
    private final FabricService fabricService;
    private final UserInfoService userInfoService;
    private final JwtUtils jwtUtils;
    private final AuthenticationManager authenticationManager;
    private final FabricUserRegistration fabricUserRegistration;

    @PostMapping("/login")
    @ResponseStatus(HttpStatus.OK)
    public String login(@RequestBody LoginRequest loginRequest) {
        Authentication authentication = authenticationManager.authenticate(
                new CustomUsernamePasswordAuthenticationToken(
                        loginRequest.getUsername(),
                        loginRequest.getPassword(),
                        null,
                        loginRequest.getMspId()
                )
        );

        if (authentication.isAuthenticated()) {
            System.out.println("User " + loginRequest.getUsername() + " is authenticated");
            return jwtUtils.generateTokenFromUserDetails(loginRequest.getUsername(), loginRequest.getMspId());
        } else {
            throw new UsernameNotFoundException("Invalid user request");
        }
    }

    @PostMapping("/submit")
    public String submitTransaction(HttpServletRequest request,
                                    @RequestParam String channelName,
                                    @RequestParam String chaincodeName,
                                    @RequestParam String functionName,
                                    @RequestParam String... args){
        String jwt = jwtUtils.getJwtFromHeader(request);
        System.out.println("jwt received");
        String username = jwtUtils.getUserNameFromJwtToken(jwt);
        String mspId = jwtUtils.getMspIdFromJwtToken(jwt);
        return fabricService.submitTransaction(channelName, chaincodeName, functionName, args,username,mspId);
    }

    @GetMapping("/query")
    public String queryTransaction(HttpServletRequest request,
                                   @RequestParam String channelName,
                                    @RequestParam String chaincodeName,
                                    @RequestParam String functionName,
                                    @RequestParam String... args){
        String jwt = jwtUtils.getJwtFromHeader(request);
        String username = jwtUtils.getUserNameFromJwtToken(jwt);
        String mspId = jwtUtils.getMspIdFromJwtToken(jwt);
        return fabricService.evaluateTransaction(channelName, chaincodeName, functionName, args,username,mspId);
    }


    @PostMapping("/enrollAdmin")
    public String enrollAdmin() {
        return "Admin enrolled successfully";
    }


    @PostMapping("/register")
    public String enrollUser(HttpServletRequest request,
                             @RequestParam String username,
                             @RequestParam String password,
                             @RequestParam(value = "file",required = false ) MultipartFile file) {

        String jwt = jwtUtils.getJwtFromHeader(request);
        String mspId = jwtUtils.getMspIdFromJwtToken(jwt);
        String id = jwtUtils.getUserNameFromJwtToken(jwt);
        System.out.println("Register controller");
        if(file!=null){
            if(!"Org2MSP".equals(mspId) )
                return "Only Patient Admin can Upload pdf";
            try {
                ObjectMapper objectMapper = new ObjectMapper();
                EhrDocument ehrDocument = objectMapper.readValue(file.getInputStream(),EhrDocument.class);
                ehrService.addEhrDocument(username, ehrDocument);
            } catch (Exception e) {
                System.err.println("Error during PDF upload: " + e.getMessage());
                return "Error During uploading pdf";
            }
        }

        System.out.println("Received");
        if(userInfoService.addUser(UserEntity.builder()
                .username(username)
                .password(password)
                .mspId(mspId)
                .build()) && fabricUserRegistration.addUser(username,password,mspId))
            return "User registered successfully";
        else
            return  "User registration failed";
    }
}

File: backend/src/controllers/FabricController.ts
import { Router, Request, Response } from 'express';
import { authMiddleware, AuthRequest } from '../middleware/authMiddleware';
import { jwtUtils } from '../middleware/jwtUtils';
import { fabricService } from '../services/FabricService';
import { ehrService } from '../services/EhrService';
import { userInfoService } from '../services/UserInfoService';
import { fabricUserRegistration } from '../services/FabricUserRegistration';
import { LoginRequest } from '../types/LoginRequest';
import { EhrDocument } from '../types/EhrDocument';
import multer from 'multer';

const router = Router();
const upload = multer({ storage: multer.memoryStorage() });

/**
 * POST /fabric/login
 * User authentication
 */
router.post('/login', async (req: Request, res: Response) => {
    try {
        const { username, password, mspId }: LoginRequest = req.body;

        if (!username || !password || !mspId) {
            res.status(400).json({ message: 'Username, password, and mspId are required' });
            return;
        }

        // Find user
        const user = await userInfoService.findByUsername(username, mspId);

        if (!user) {
            res.status(401).json({ message: 'Invalid credentials' });
            return;
        }

        // Verify password
        const isPasswordValid = await userInfoService.verifyPassword(user, password);

        if (!isPasswordValid) {
            res.status(401).json({ message: 'Invalid credentials' });
            return;
        }

        // Generate JWT
        const token = jwtUtils.generateToken(username, mspId);

        console.log(`User ${username} authenticated successfully`);
        res.status(200).json({ token });
    } catch (error) {
        console.error('Login error:', error);
        res.status(500).json({ message: 'Login failed' });
    }
});

/**
 * POST /fabric/submit
 * Submit transaction to blockchain
 */
router.post('/submit', authMiddleware, async (req: AuthRequest, res: Response) => {
    try {
        const { channelName, chaincodeName, functionName, args } = req.body;

        if (!channelName || !chaincodeName || !functionName) {
            res.status(400).json({ message: 'Missing required parameters' });
            return;
        }

        const username = req.user!.username;
        const mspId = req.user!.mspId;

        console.log('JWT received, submitting transaction');

        const result = await fabricService.submitTransaction(
            channelName,
            chaincodeName,
            functionName,
            args || [],
            username,
            mspId
        );

        res.status(200).send(result);
    } catch (error) {
        console.error('Submit transaction error:', error);
        res.status(500).json({ message: 'Transaction submission failed' });
    }
});

/**
 * GET /fabric/query
 * Query blockchain
 */
router.get('/query', authMiddleware, async (req: AuthRequest, res: Response) => {
    try {
        const { channelName, chaincodeName, functionName } = req.query;
        const args = req.query.args ? (Array.isArray(req.query.args) ? req.query.args : [req.query.args]) : [];

        if (!channelName || !chaincodeName || !functionName) {
            res.status(400).json({ message: 'Missing required parameters' });
            return;
        }

        const username = req.user!.username;
        const mspId = req.user!.mspId;

        const result = await fabricService.evaluateTransaction(
            channelName as string,
            chaincodeName as string,
            functionName as string,
            args as string[],
            username,
            mspId
        );

        res.status(200).send(result);
    } catch (error) {
        console.error('Query transaction error:', error);
        res.status(500).json({ message: 'Query failed' });
    }
});

/**
 * POST /fabric/enrollAdmin
 * Enroll admin (placeholder)
 */
router.post('/enrollAdmin', async (_req: Request, res: Response) => {
    res.status(200).send('Admin enrolled successfully');
});

/**
 * POST /fabric/register
 * Register new user
 */
router.post('/register', authMiddleware, upload.single('file'), async (req: AuthRequest, res: Response) => {
    try {
        const { username, password } = req.body;
        const mspId = req.user!.mspId;

        if (!username || !password) {
            res.status(400).json({ message: 'Username and password are required' });
            return;
        }

        console.log('Register controller');

        // Handle EHR document upload if file is provided
        if (req.file) {
            if (mspId !== 'Org2MSP') {
                res.status(403).send('Only Patient Admin can upload PDF');
                return;
            }

            try {
                const ehrDocument: EhrDocument = JSON.parse(req.file.buffer.toString('utf8'));
                await ehrService.addEhrDocument(username, ehrDocument);
            } catch (error) {
                console.error('Error during PDF upload:', error);
                res.status(500).send('Error during uploading PDF');
                return;
            }
        }

        console.log('Received registration request');

        // Add user to database
        const userAdded = await userInfoService.addUser({ username, password, mspId });

        // Add user to Fabric network
        const fabricUserAdded = await fabricUserRegistration.addUser(username, password, mspId);

        if (userAdded && fabricUserAdded) {
            res.status(200).send('User registered successfully');
        } else {
            res.status(500).send('User registration failed');
        }
    } catch (error) {
        console.error('Registration error:', error);
        res.status(500).json({ message: 'Registration failed' });
    }
});

export default router;
